clients/src/main/java/org/apache/kafka/clients/telemetry/DefaultClientTelemetry.java

```diff
@@ -398,6 +404,46 @@ else if (state == TelemetryState.terminating_push_in_progress)
             throw new IllegalTelemetryStateException(String.format("Could not transition state after successful push telemetry from state %s", state));
     }
 
+    /**
+     * Examine the response data and handle different error code accordingly.
+     *  - Authorization Failed: Retry 30min later
+     *  - Invalid Record: Retry 5min later
+     *  - UnknownSubscription or Unsupported Compression: Retry
+     *  After a new subscription is created, the current state is transitioned to push_needed to wait for another push.
+     *
+     * @param data pushTelemetryResponseData
+     * @throws IllegalTelemetryStateException when subscription is mull
+     */
+    private void handlePushTelemetryResponseDataErrors(final PushTelemetryResponseData data) {
+        TelemetrySubscription currentSubscription = subscription().orElseThrow(
+        () -> new IllegalTelemetryStateException(String.format("Could not transition state after successful push telemetry from state %s", state)));
+
+        // We might want to wait and retry or retry after some failures are received
+        if (isAuthorizationFailedError(data.errorCode())) {
+            final long retryMs = 30 * 60 * 1000;
+            log.warn("Error code: {}. Reason: Client is permitted to send metrics.  Retry automatically in {}ms.", data.errorCode(), retryMs);
+            setSubscription(currentSubscription.alterPushIntervalMs(retryMs));
+        } else if (data.errorCode() == Errors.INVALID_RECORD.code()) {
+            final long retryMs = 5 * 60 * 1000;
+            log.warn("Error code: {}.  Reason: Broker failed to decode or validate the client’s encoded metrics.  Retry automatically in {}ms", data.errorCode(), retryMs);
+            setSubscription(currentSubscription.alterPushIntervalMs(retryMs));
+        } else if (data.errorCode() == -1 ||
+                data.errorCode() == Errors.UNSUPPORTED_COMPRESSION_TYPE.code()) {
+            setSubscription(currentSubscription.alterPushIntervalMs(0));
```

Comment:
Do we want any logging here for the 'unsupported compression' error?

Reply:
It would be a good idea, but it is not specified in the kip, so I didn't add it there.